components/external_storage/src/util.rs

```diff
@@ -0,0 +1,69 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+use bytes::Bytes;
+use futures::stream::{self, Stream};
+use futures_io::AsyncRead;
+use std::{
+    future::Future,
+    io, iter,
+    marker::Unpin,
+    pin::Pin,
+    sync::Mutex,
+    task::{Context, Poll},
+};
+
+/// Wrapper of an `AsyncRead` instance, exposed as a `Sync` `Stream` of `Bytes`.
+pub struct AsyncReadAsSyncStreamOfBytes<R> {
+    // we need this Mutex to ensure the type is Sync (provided R is Send).
+    // this is because rocksdb::SequentialFile is *not* Sync
+    // (according to the documentation it cannot be Sync either,
+    // requiring "external synchronization".)
+    reader: Mutex<R>,
+    // we use this member to ensure every call to `poll_next()` reuse the same
+    // buffer.
```

Comment:
why reuse the buffer is needed? for performance?

Reply:
And noob question: why the struct is `Sync` since the buf is not `Sync`?

Reply:
> why reuse the buffer is needed? for performance?

The buffer was reused initially because of #6675 (which later I found a simpler fix and the whole module was scraped out from 6675). I'm too lazy to re-implement this by other means 🙃.

(The other alternative is allocate an uninitialized vector every poll and release its ownership at the end.)

> And noob question: why the struct is Sync since the buf is not Sync?

Um, the `buf` is Sync, `Vec<u8>` is Sync. 

`T: Sync` means `&T: Send` i.e. the shared reference can be safely reused among threads, which is true for *immutable* types like `Vec<u8>`.

(The more practical reason is that it is Sync because the compiler accepts that it is Sync 😋)